### DISCUSSION TITLE
Oberon SET concept

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#6862] [Q&A] [answered]
Is there anything in C# (already exists or under consideration) that corresponds to the [SET](https://people.inf.ethz.ch/wirth/Oberon/SETs.pdf) as defined in the Oberon (and earlier Wirth) languages?

**Comment by HaloFour** [comment, marked as answer]
The BCL offers [BitArray](https://learn.microsoft.com/en-us/dotnet/api/system.collections.bitarray?view=net-7.0) and [BitVector32](https://learn.microsoft.com/en-us/dotnet/api/system.collections.specialized.bitvector32?view=net-7.0) data types.  C# has no special data types for this, it only offers integral data types and bitwise operators on them.

**Reply by The-Futurist** [reply]
> The BCL offers [BitArray](https://learn.microsoft.com/en-us/dotnet/api/system.collections.bitarray?view=net-7.0) and [BitVector32](https://learn.microsoft.com/en-us/dotnet/api/system.collections.specialized.bitvector32?view=net-7.0) data types. C# has no special data types for this, it only offers integral data types and bitwise operators on them.

That is very similar, implemented in a compact manner, thanks. Do you now where it gets used? what kind of problems leverage it?

**Reply by CyrusNajmabadi** [reply]
I used one in our implementation of our bloom filter, to store booleans at particular offsets for hash hits.  Compiler uses something very similar to work with control-flow blocks (like if they are reachable or not), definite-assignment tracking, nullability-analysis, etc. .  They are good for problems where you have a lot of indices for things you want to store bools with, and you'd like the space for that to be compact.

**Comment by The-Futurist** [comment]
Thanks.
